lib/spack/spack/cmd/config.py

```diff
@@ -135,11 +164,150 @@ def config_list(args):
     print(' '.join(list(spack.config.section_schemas)))
 
 
+def config_add(args):
+    """Add the given configuration to the specified config scope
+
+    This is a stateful operation that edits the config files under the hood"""
+    if not (args.file or args.value):
+        tty.error("No changes requested. Specify a file or value.")
+        setup_parser.add_parser.print_help()
+        exit(1)
+
+    scope, section = _get_scope_and_section(args)
+
+    # Updates from file
+    if args.file:
+        # Get file as config dict
+        with open(args.file, 'r') as f:
+            data = syaml.load_config(f)
+
+        if 'env' in data or 'spack' in data:
+            # Validate as an environment
+            # use config_dict to remove top-level key
+            spack.environment.validate(data, 'spack.yaml')
+            config_dict = spack.environment.config_dict(data)
+        else:
+            # Separate config file
+            # Use schema according to first key (they each only accept one key)
+            # Do not remove top key
+            spack.config.validate(
+                data, spack.config.section_schemas[next(iter(data))])
+            config_dict = data
+        # update all sections from config dict
+        # We have to iterate on keys to keep overrides from the file
+        for section in config_dict.keys():
+            if section in spack.config.section_schemas.keys():
+                # Special handling for compiler scope difference
+                # Has to be handled after we choose a section
+                if scope is None:
+                    if section == 'compilers':
+                        scope = spack.config.default_modify_scope()
+                    else:
+                        scope = spack.config.default_modify_scope(False)
+
+                value = config_dict[section]
+                existing = spack.config.get(section, scope=scope)
+                new = spack.config.merge_yaml(existing, value)
+
+                if re.match(r'env.*', scope or ''):
+                    e = ev.get_env(args, 'config add')
+                    e.set_config(section, new)
```

Comment:
I think we should instead make `spack.config.set()` work with `SingleFileScope` -- is that possible here?  Seems like this would be much cleaner with a small refactor in `spack.config`, and you'd only have one case to deal with for setting things.